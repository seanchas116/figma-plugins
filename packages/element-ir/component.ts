export interface Component {
  key?: string;
  element: Element;
  propertyDefinitions: PropertyDefinition[];
}

export interface PropertyDefinition {
  key: string;
  type: PropertyType;
}

export type PropertyType =
  | {
      type: "boolean";
    }
  | {
      type: "string";
    }
  | {
      type: "enum";
      options: string[];
    }
  | {
      type: "instance";
    };
