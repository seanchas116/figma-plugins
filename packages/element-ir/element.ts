import {
  FrameStyle,
  ImageStyle,
  InstanceStyle,
  SVGStyle,
  TextStyle,
} from "./style";

export interface CommonProps {
  id: string;
  name: string;
  propertyRef: {
    visible?: string;
    content?: string;
    component?: string;
  };
}

export interface FrameElement extends CommonProps {
  type: "frame";
  style: FrameStyle;
  children: Element[];
}

export interface TextElement extends CommonProps {
  type: "text";
  style: TextStyle;
  content: string; // TODO: spans
}

export interface ImageElement extends CommonProps {
  type: "image";
  style: ImageStyle;
  imageID: string;
}

export interface SVGElement extends CommonProps {
  type: "svg";
  style: SVGStyle;
  svg: string;
}

export interface InstanceElement extends CommonProps {
  type: "instance";
  componentKey: string;
  properties: Record<string, any>;
  // overrides are not supported (instances are inlined if any overrides are present)
  style: InstanceStyle;
}

export type Element =
  | FrameElement
  | TextElement
  | ImageElement
  | SVGElement
  | InstanceElement;
