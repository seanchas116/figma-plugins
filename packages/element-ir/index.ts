export * from "./style";
export * from "./element";
export * from "./component";
