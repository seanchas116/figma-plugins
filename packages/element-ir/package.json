{
  "name": "@uimix/element-ir",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "typings": "index.d.ts",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "author": "",
  "license": "MIT"
}
