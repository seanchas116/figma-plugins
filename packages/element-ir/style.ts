export interface Color {
  r: number;
  g: number;
  b: number;
  a: number;
}

export interface SolidPaint {
  type: "solid";
  color: Color;
}

export interface Vector {
  x: number;
  y: number;
}

export interface ColorStop {
  position: number;
  color: Color;
}

export interface GradientPaint {
  type: "gradientLinear" | "gradientRadial" | "gradientAngular";
  handles: [Vector, Vector, Vector];
  stops: ColorStop[];
}

export interface ImagePaint {
  type: "image";
  imageID: string;
  size: "fill" | "contain" | "cover";
}

export type Paint = SolidPaint | GradientPaint | ImagePaint;

export interface DimensionStyleMixin {
  position: "absolute" | "relative";
  display: "flex" | "none";
  x:
    | { left: number }
    | { right: number }
    | { centerRatio: number }
    | { left: number; right: number };
  y:
    | { top: number }
    | { bottom: number }
    | { centerRatio: number }
    | { top: number; bottom: number };
  width: number | "fit-content" | "stretch";
  height: number | "fit-content" | "stretch";
}

export interface RectangleFillBorderStyleMixin {
  borderTopLeftRadius: number;
  borderTopRightRadius: number;
  borderBottomLeftRadius: number;
  borderBottomRightRadius: number;

  border: Paint[];
  borderTopWidth: number;
  borderRightWidth: number;
  borderBottomWidth: number;
  borderLeftWidth: number;

  background: Paint[];
}

export interface FrameStyleMixin {
  overflow: "hidden" | "visible"; // TODO: scroll
  flexDirection: "row" | "column";
  gap: number;
  paddingTop: number;
  paddingBottom: number;
  paddingLeft: number;
  paddingRight: number;
  alignItems: "center" | "flex-start" | "flex-end" | "baseline";
  justifyContent: "center" | "flex-start" | "flex-end" | "space-between";
}

export interface TextSpanStyleMixin {
  fontFamily: string;
  fontStyle: "normal" | "italic";
  fontWeight: number;
  fontSize: number;
  lineHeight: number | "normal";
  letterSpacing: number; // em;
  color: Paint[];
}

export interface TextStyleMixin {
  textAlign: "left" | "center" | "right" | "justify";
  justifyContent: "center" | "flex-start" | "flex-end"; // assuming a text element becomes a vertical flexbox
}

export interface ImageStyleMixin {
  objectFit: "fill" | "contain" | "cover";
}

export interface FrameStyle
  extends DimensionStyleMixin,
    RectangleFillBorderStyleMixin,
    FrameStyleMixin {}

export interface TextStyle
  extends DimensionStyleMixin,
    TextSpanStyleMixin,
    TextStyleMixin {}

export interface SVGStyle extends DimensionStyleMixin {}

export interface ImageStyle
  extends DimensionStyleMixin,
    RectangleFillBorderStyleMixin,
    ImageStyleMixin {}
