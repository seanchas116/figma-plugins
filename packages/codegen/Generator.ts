import { Component, Element } from "@uimix/element-ir";
import { camelCase, capitalize } from "lodash-es";
import * as svgParser from "svg-parser";
import { InlineStyleGenerator } from "./style/InlineStyleGenerator";
import { IStyleGenerator } from "./style/IStyleGenerator";

export interface GeneratorOptions {
  jsx?: boolean;
  components?: Record<string, Component>;
  styleGenerator?: IStyleGenerator;
}

export class Generator {
  constructor(options: GeneratorOptions = {}) {
    this.options = {
      jsx: options.jsx ?? false,
      components: new Map(Object.entries(options.components ?? {})),
    };
    this.styleGenerator = options.styleGenerator ?? new InlineStyleGenerator();
  }

  private options: {
    jsx: boolean;
    components: Map<string, Component>;
  };
  readonly styleGenerator: IStyleGenerator;

  private generateTag(
    tagName: string,
    props: Record<string, any>,
    children: string[] = []
  ): string[] {
    const propsStr = Object.entries(props)
      .map(([key, value]) =>
        value ? `${key}={${JSON.stringify(value)}}` : key
      )
      .join(" ");
    return [`<${tagName} ${propsStr}>`, ...children, `</${tagName}>`];
  }

  generateElement(element: Element): string[] {
    switch (element.type) {
      case "instance": {
        const component = this.options.components.get(element.componentKey);
        if (component) {
          const props: Record<string, string> = {};
          for (const [key, value] of Object.entries(element.properties)) {
            const codeName = camelCase(key.split("#")[0]);
            props[codeName] = value;
          }
          console.log(props);
          // TODO: generate unique name
          const name = capitalize(camelCase(component.element.name));
          return this.generateTag(name, props);
        }
        console.error('Component not found: "' + element.componentKey + '"');
        return [];
      }
      case "frame": {
        return this.generateTag(
          "div",
          {
            style: this.styleGenerator.frameCSS(element.style),
          },
          element.children.flatMap((e) => this.generateElement(e))
        );
      }
      case "image": {
        return this.generateTag("img", {
          style: this.styleGenerator.imageCSS(element.style),
        });
      }
      case "svg": {
        const root = svgParser.parse(element.svg);
        const svgElem = root.children[0];
        if (svgElem.type !== "element") {
          throw new Error("Expected element type");
        }

        const svgChildren = element.svg
          .trim()
          .replace(/^<svg[^>]*>/, "")
          .replace(/<\/svg>$/, "");

        const properties: Record<string, any> = {
          ...svgElem.properties,
          style: {
            ...this.styleGenerator.svgCSS(element.style),
          },
        };
        delete properties.xmlns;

        return this.generateTag("svg", properties, [svgChildren]);
      }
      case "text": {
        return this.generateTag(
          "div",
          {
            style: this.styleGenerator.textCSS(element.style),
          },
          [element.content]
        );
      }
    }
  }

  generateComponent(component: Component) {
    // TODO: generate unique name
    const name = capitalize(camelCase(component.element.name));

    const element = {
      ...component.element,
      style: {
        ...component.element.style,
        position: "relative",
        x: { left: 0 },
        y: { top: 0 },
      },
    };

    return [
      `export function ${name}() { return `,
      ...this.generateElement(element as Element),
      `}`,
    ];
  }
}
