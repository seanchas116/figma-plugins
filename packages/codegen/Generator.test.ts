import { describe, expect, it } from "vitest";
import { Generator } from "./Generator";
import { Component } from "@uimix/element-ir";

import _components from "./__fixtures__/components.json";
const components: Component[] = _components as any;

describe("Generator", () => {
  for (const style of ["tailwind", "inline"] as const) {
    it(`should emit ${style}`, () => {
      const generator = new Generator({
        components,
        style,
      });

      expect(generator.generateProject()).toMatchSnapshot();
    });
  }
});
