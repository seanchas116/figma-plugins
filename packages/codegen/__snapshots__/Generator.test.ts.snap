// Vitest Snapshot v1

exports[`Generator > should work 1`] = `
[
  {
    "content": "export function Button(props) {
  return (
    <div
      {...props}
      className={twMerge(
        \\"flex relative flex-none self-auto w-fit h-fit border-transparent bg-blue-500 rounded-tl rounded-tr rounded-br rounded-bl border-t-0 border-r-0 border-b-0 border-l-0 overflow-hidden flex-row gap-1.5 pt-1.5 pr-3 pb-1.5 pl-3 items-center justify-center\\",
        props.className
      )}
    >
      <div className=\\"flex relative flex-none self-auto w-fit h-fit font-['Inter'] text-xs font-medium not-italic leading-normal tracking-normal text-white text-left flex-col justify-start\\">
        {props.text}
      </div>
      {props.showChevron && (
        <div className=\\"flex relative flex-none self-auto w-2 h-2 border-transparent bg-transparent rounded-tl-none rounded-tr-none rounded-br-none rounded-bl-none border-t-0 border-r-0 border-b-0 border-l-0 overflow-visible flex-row gap-0 pt-0 pr-0 pb-0 pl-0 items-start justify-start\\">
          <svg
            width={5}
            height={7}
            viewBox={\\"0 0 5 7\\"}
            fill={\\"none\\"}
            className=\\"flex absolute left-[0.1302083283662796rem] top-[0.0416666679084301rem] flex-none self-auto w-[0.24531249701976776rem] h-[0.4166666567325592rem]\\"
          >
            <path
              d=\\"M0.674995 6.83334L0.0833282 6.24167L2.825 3.5L0.0833282 0.758338L0.674995 0.166672L4.00833 3.5L0.674995 6.83334Z\\"
              fill=\\"white\\"
            />
          </svg>
        </div>
      )}
    </div>
  );
}
",
    "filePath": "Button.js",
  },
  {
    "content": "import { Button } from \\"./Button.js\\";

export function Form2(props) {
  return (
    <div
      {...props}
      className={twMerge(
        \\"flex relative flex-none self-auto w-fit h-fit border-transparent bg-white rounded-tl-none rounded-tr-none rounded-br-none rounded-bl-none border-t-0 border-r-0 border-b-0 border-l-0 overflow-hidden flex-col gap-[0.8125rem] pt-4 pr-4 pb-4 pl-4 items-start justify-start\\",
        props.className
      )}
    >
      <div className=\\"flex relative flex-none self-auto w-fit h-fit font-['Inter'] text-xs font-normal not-italic leading-normal tracking-normal text-black text-left flex-col justify-start\\">
        Some Form (stretch)
      </div>
      <Button
        showChevron={false}
        text={\\"OK\\"}
        className=\\"self-stretch w-auto\\"
      ></Button>
    </div>
  );
}
",
    "filePath": "Form2.js",
  },
  {
    "content": "import { Button } from \\"./Button.js\\";

export function Form(props) {
  return (
    <div
      {...props}
      className={twMerge(
        \\"flex relative flex-none self-auto w-fit h-fit border-transparent bg-white rounded-tl-none rounded-tr-none rounded-br-none rounded-bl-none border-t-0 border-r-0 border-b-0 border-l-0 overflow-hidden flex-col gap-[0.8125rem] pt-4 pr-4 pb-4 pl-4 items-start justify-start\\",
        props.className
      )}
    >
      <div className=\\"flex relative flex-none self-auto w-fit h-fit font-['Inter'] text-xs font-normal not-italic leading-normal tracking-normal text-black text-left flex-col justify-start\\">
        Some Form
      </div>
      <Button showChevron={true} text={\\"OK\\"} className=\\"\\"></Button>
    </div>
  );
}
",
    "filePath": "Form.js",
  },
  {
    "content": "export function Frame1(props) {
  return (
    <div
      {...props}
      className={twMerge(
        \\"flex relative flex-none self-auto w-[21.5rem] h-[5.8125rem] border-transparent bg-white rounded-tl-none rounded-tr-none rounded-br-none rounded-bl-none border-t-0 border-r-0 border-b-0 border-l-0 overflow-hidden flex-col gap-0 pt-0 pr-0 pb-0 pl-0 items-start justify-start\\",
        props.className
      )}
    >
      <svg
        width={128}
        height={64}
        viewBox={\\"0 0 128 64\\"}
        fill={\\"none\\"}
        className=\\"flex absolute left-2 top-5 flex-none self-auto w-32 h-16\\"
      >
        <rect width=\\"128\\" height=\\"64\\" fill=\\"#D9D9D9\\" />
      </svg>
      <svg
        width={128}
        height={64}
        viewBox={\\"0 0 128 64\\"}
        fill={\\"none\\"}
        className=\\"flex absolute left-[9.5625rem] top-2 flex-none self-auto w-32 h-16\\"
      >
        <rect width=\\"128\\" height=\\"64\\" fill=\\"#FF0000\\" />
      </svg>
    </div>
  );
}
",
    "filePath": "Frame1.js",
  },
]
`;
